cryptarchia/cryptarchia.py

```diff
@@ -93,20 +143,57 @@ def block_position(self, block: BlockHeader) -> int:
                 return i
 
 
+@dataclass
+class LedgerState:
+    """
+    A snapshot of the ledger state up to some block
+    """
+
+    block: Id = None
+    nonce: Id = None
+    total_stake: int = None
+    commitments: set[Id] = field(default_factory=set)  # set of commitments
+    nullifiers: set[Id] = field(default_factory=set)  # set of nullified
+
+    def verify_committed(self, commitment: Id) -> bool:
+        return commitment in self.commitments
+
+    def verify_unspent(self, nullifier: Id) -> bool:
+        return nullifier not in self.nullifiers
+
+    def apply(self, block: BlockHeader):
+        assert block.parent == self.block
+        self.block = block.id()
+        self.nullifiers.add(block.leader_proof.nullifier)
```

Comment:
the new commitment will be added in the next PR right?

Reply:
yes, that's correct, leader proofs will generate a new commitment